Python/pythonrun.c

```diff
@@ -280,11 +280,44 @@ PyRun_InteractiveOneObjectEx(FILE *fp, PyObject *filename,
     PyObject *main_dict = PyModule_GetDict(main_module);  // borrowed ref
 
     PyObject *res = run_mod(mod, filename, main_dict, main_dict, flags, arena, interactive_src, 1);
+    Py_INCREF(interactive_src);
     _PyArena_Free(arena);
     Py_DECREF(main_module);
     if (res == NULL) {
+        PyThreadState *tstate = _PyThreadState_GET();
+        PyObject *exc = _PyErr_GetRaisedException(tstate);
+        if (PyType_IsSubtype(Py_TYPE(exc),
+                             (PyTypeObject *) PyExc_SyntaxError))
+        {
+            /* fix "text" attribute */
+            assert(interactive_src != NULL);
+            PyObject *xs = PyUnicode_Splitlines(interactive_src, 1);
+            if (xs == NULL) {
+                goto err;
+            }
+            PyObject *ln = PyObject_GetAttr(exc, &_Py_ID(lineno));
+            if (ln == NULL) {
+                Py_DECREF(xs);
+                goto err;
+            }
+            int n = PyLong_AsInt(ln);
+            Py_DECREF(ln);
+            if (n <= 0 || n > PyList_GET_SIZE(xs)) {
+                Py_DECREF(xs);
+                goto err;
+            }
+            PyObject *line = PyList_GET_ITEM(xs, n - 1);
+            if (PyObject_SetAttr(exc, &_Py_ID(text), line) == -1) {
+                _PyErr_Clear(tstate);
+            }
+            Py_DECREF(xs);
+        }
+err:
```

Comment:
I would use perhaps `error` or `abort` for the label name (I think there are more occurrences of `error` (2k) than `err` (<100) in the library in general).